nltk/stem/cistem.py

```diff
@@ -0,0 +1,216 @@
+        self._strip_nd = re.compile(r"nd$")
+        self._strip_t = re.compile(r"t$")
+        self._strip_esn = re.compile(r"[esn]$")
+
+    def stem(self, word):
+        """
+        This method takes the word to be stemmed and a boolean specifiying if
+        case-insensitive stemming should be used and returns the stemmed word. If only
+        the word is passed to the method or the second parameter is 0, normal
+        case-sensitive stemming is used, if the second parameter is 1,
+        case-insensitive stemming is used.
+
+        Case sensitivity improves performance only if words in the text may be
+        incorrectly upper case. For all-lowercase and correctly cased text, best
+        performance is achieved byusing the case-sensitive version.
+
+        :param word: the word that is to be stemmed
+        :type word: unicode
+        :return word: the stemmed word
+        :rtype: unicode
+
+        >>> from nltk.stem.cistem import Cistem
+        >>> stemmer = Cistem()
+        >>> s1 = "Speicherbehältern"
+        >>> stemmer.stem(s1)
+        'speicherbehalt'
+        >>> s2 = "Grenzpostens"
+        >>> stemmer.stem(s2)
+        'grenzpost'
+        >>> s3 = "Ausgefeiltere"
+        >>> stemmer.stem(s3)
+        'ausgefeilt'
+        >>> stemmer = Cistem(True)
+        >>> stemmer.stem(s1)
+        'speicherbehal'
+        >>> stemmer.stem(s2)
+        'grenzpo'
+        >>> stemmer.stem(s3)
+        'ausgefeil'
+        """
+        if len(word) == 0:
+            return word
+
+        upper = word[0].isupper()
+        word = word.lower()
+
+        word = word.replace("ü", "u")
+        word = word.replace("ö", "o")
+        word = word.replace("ä", "a")
+        word = word.replace("ß", "ss")
+
+        word = self._strip_ge.sub(r"\1", word)
+        word = word.replace("sch", "$")
+        word = word.replace("ei", "%")
+        word = word.replace("ie", "&")
+        word = self._repl_xx.sub(r"\1*", word)
+
+        while len(word) > 3:
+            if len(word) > 5:
+                (word, success) = self._strip_emr.subn("", word)
+                if success != 0:
+                    continue
+
+                (word, success) = self._strip_nd.subn("", word)
+                if success != 0:
+                    continue
+
+            if not upper or self._case_insensitive:
+                (word, success) = self._strip_t.subn("", word)
+                if success != 0:
+                    continue
+
+            (word, success) = self._strip_esn.subn("", word)
+            if success != 0:
+                continue
+            else:
+                break
+
+        word = self._repl_xx_back.sub(r"\1\1", word)
+        word = word.replace("%", "ei")
+        word = word.replace("&", "ie")
+        word = word.replace("$", "sch")
+
+        return word
+
+
+    def segment(self, word):
+        """
+        This method works very similarly to stem. The difference is that in
```

Comment:
NB. Not essential but note that Sphinx lets you refer to other methods in a way that generates a link in the HTML docs (plenty of examples in NLTK code)